client/src/contexts/AuthContext.js

```diff
@@ -0,0 +1,44 @@
+import React, { createContext, useState } from "react";
+import PropTypes from "prop-types";
+import axios from "axios";
+
+export const AuthContext = createContext();
+
+AuthContextProvider.propTypes = {
+  children: PropTypes.node,
+};
+
+export function AuthContextProvider({ children }) {
+  const [token, setToken] = useState();
+  const [user, setUser] = useState();
+
+  const loginWithCode = (code) => {
+    if (token) {
+      return;
+    }
+    axios
+      .post("/login/oauth/github/token", { code })
+      .then((res) => res.data.access_token)
+      .then((token) => {
+        setToken(token);
+        return axios.get("https://api.github.com/user", {
+          headers: {
+            Authorization: `Bearer ${token}`,
+          },
+        });
+      })
+      .then((res) => setUser(res.data))
+      .catch(console.log);
+  };
```

Comment:
Consider using `useCallback` here for a memoized function.

```suggestion
  const loginWithCode = useCallback((code) => {
    if (token) {
      return;
    }
    axios
      .post("/login/oauth/github/token", { code })
      .then((res) => res.data.access_token)
      .then((token) => {
        setToken(token);
        return axios.get("https://api.github.com/user", {
          headers: {
            Authorization: `Bearer ${token}`,
          },
        });
      })
      .then((res) => setUser(res.data))
      .catch(console.log);
  }, [token]);
```

https://reactjs.org/docs/hooks-reference.html#usecallback